src/semantics.js

```diff
@@ -160,73 +173,107 @@ var semicolonCmdNames = [
   },
   ControlStruct: function(assign, _sc1, id, binop, val, _sc2, update) {
-    return assign.toJS(0) + ';' + id.sourceString + binop.toJS(0) + val.toJS(0) +
+    return assign.toJS(0, this.args.ctx) + ';' + id.sourceString + binop.toJS(0, this.args.ctx) + val.toJS(0, this.args.ctx) +
       ';' + update.sourceString;
   },
   ForCommand_for_each: function(_for, id, _in, call, _sc, _do, _s, cmd2, done) {
-    var mycall = call.toJS(this.args.indent).replace(/\.replace\(.*\)/, '');
+    var mycall = call.toJS(this.args.indent, this.args.ctx).replace(/\.replace\(.*\)/, '');
     return mycall + '.forEach(function (' + id.sourceString + ') {' +
-        nl(this.args.indent+1) + cmd2.toJS(this.args.indent+1) +
+        nl(this.args.indent+1) + cmd2.toJS(this.args.indent+1, this.args.ctx) +
         nl(this.args.indent) + '});';
   },
   WhileCommand: function(_w, _s, cond, _sc, _do, _s2, cmd, done) {
-    return 'while (' + cond.toJS(this.args.indent) + ') {' +
-        nl(this.args.indent+1) + cmd.toJS(this.args.indent+1) +
-        done.toJS(this.args.indent);
+    return 'while (' + cond.toJS(this.args.indent, this.args.ctx) + ') {' +
+        nl(this.args.indent+1) + cmd.toJS(this.args.indent+1, this.args.ctx) +
+        done.toJS(this.args.indent, this.args.ctx);
   },
   Done: function(_sc, _) {
     return nl(this.args.indent) + '}';
   },
   FunctionDecl: function(_fun, _sp1, id, _paren, _sp2, block) {
-    var idStr = id.toJS(0);
+    var idStr = id.toJS(0, this.args.ctx);
     allFunctions[idStr] = true;
 
     inFunctionBody = true;
-    var blockString = block.toJS(this.args.indent);
+    var blockString = block.toJS(this.args.indent, this.args.ctx);
     inFunctionBody = false;
 
     return 'function ' + idStr + '(..._$args) ' + blockString;
   },
+  CaseCommand: function(_case, expr, _in, _ws, cases, _ws2, _esac){
+    var varName = "tmp" + Date.now();
+    var indent = this.args.indent;
+
+    // If we can scan the children to see if there are glob patterns
+    // ahead of time, we can set hasGlob to false and generate cleaner code.
+    var hasGlob = true;
+
+    if (hasGlob){
+      return nl(indent) + "var " + varName + " = " +
+      expr.toJS(0,this.args.ctx) + ";" + nl(indent) +
+      "switch (" + varName + ") {" + nl(indent + 1) +
+      cases.toJS(indent + 1, Object.assign(this.args.ctx,
+        {caseVar:varName})).join(nl(indent + 1)) +
+      nl(indent) + "}";
+    } else {
+      return nl(indent) +
+      "switch (" + expr.toJS(0,this.args.ctx) + ") {" + nl(indent + 1) +
+      cases.toJS(indent + 1, this.args.ctx).join(nl(indent + 1)) +
+      nl(indent) + "}";
+    }
+  },
+  CaseCase: function(opts, _par, _ws, cmds, _ws2, _semisemi, comment){
+    var varName = this.args.ctx.caseVar;
+    var comment = comment.toJS(this.args.indent, this.args.ctx);
+    return opts.toJS(0, this.args.ctx)
+        .map(s => s.substr(1, s.length - 2))
+        .map(s => isGlobString(s) ? ("case " + s + ":")
```

Comment:
Could you rewrite this as `function () { }`? I'd like to keep compatibility with iOS safari 9.3 for a bit longer (at least until I upgrade my phone to iOS 10 :laughing:). Not a big concern, I can fix it up after the merge myself.